repose-aggregator/services/atom-feed/impl/src/main/scala/org/openrepose/nodeservice/atomfeed/impl/AtomEntryStreamBuilder.scala

```diff
@@ -0,0 +1,105 @@
+    * making a request to the Atom feed, assign the result of this function to a lazy variable like so:
+    * lazy val myFeed = build("http://my.feed.url", MyAuthenticatedRequestFactory)
+    *
+    * @param baseFeedUrl a static locator pointing to the "head" of an Atom feed (e.g., the subscription document)
+    * @param authenticator a URL processor which authenticates connections
+    * @param authenticationTimeLimit a maximum [[Duration]] to wait on authentication before considering authentication
+    *                                a failure
+    * @return a Scala [[scala.collection.immutable.Stream]] which will yield Atom entries in an Atom feed located at the
+    *         provided baseFeedUrl. The entries will be yielded in the order in which they are read from the feed XML,
+    *         from top to bottom.
+    * @throws AuthenticationException when the authenticator fails to authenticate a request
+    */
+  def build(baseFeedUrl: URL,
+            reposeVersion: String,
+            authenticator: Option[AuthenticatedRequestFactory] = None,
+            authenticationTimeLimit: Duration = 1 second): Stream[Entry] = {
+    val requestId = java.util.UUID.randomUUID().toString
+    MDC.put(TracingKey.TRACING_KEY, requestId)
+    buildR(baseFeedUrl, reposeVersion, authenticator, authenticationTimeLimit, requestId)
+  }
+
+  private def buildR(baseFeedUrl: URL,
+                     reposeVersion: String,
+                     authenticator: Option[AuthenticatedRequestFactory] = None,
+                     authenticationTimeLimit: Duration = 1 second,
+                     requestId: String): Stream[Entry] = {
+    val baseFeedConnection = baseFeedUrl.openConnection()
+
+    val authenticatedConnection = authenticator match {
+      case Some(arf) =>
+        val connectionFuture = Future(arf.authenticateRequest(baseFeedConnection, requestId, reposeVersion))
+        Option(Await.result(connectionFuture, authenticationTimeLimit))
+      case None =>
+        Some(baseFeedConnection)
+    }
+
+    authenticatedConnection match {
+      case Some(urlConnection) =>
+        val tracingHeader = TracingHeaderHelper.createTracingHeader(requestId, "1.1 Repose (Repose/" + reposeVersion + ")", None)
+        urlConnection.setRequestProperty(CommonHttpHeader.TRACE_GUID.toString, tracingHeader)
+
+        val feedInputStream = urlConnection.getInputStream
+        val feed = parser.parse[Feed](feedInputStream).getRoot
+        feedInputStream.close()
+
+        feed.getLinks.find(link => link.getRel.equals("next")) match {
+          case Some(nextPageLink) =>
+            feed.getEntries.toStream #::: buildR(nextPageLink.getResolvedHref.toURL, reposeVersion, authenticator, authenticationTimeLimit, requestId)
+          case None =>
+            feed.getEntries.toStream
+        }
+      case None =>
+        throw AuthenticationException
```

Comment:
This is the case where the AuthenticatedRequestFactory returns a null (to signify that something went wrong). If we wanted to retry, that could absolutely be done in the AuthenticatedRequestFactory itself.

In this case, the page failed to be read due to failed authentication. The high water mark is not updated, but events are potentially sent. When the schedule kicks again, authentication will be attempted and flow will continue as normal. So there is a risk of duplicate events...

I think an easy fix would be to set the high water mark earlier. Rather than waiting until the end of the stream, do it for each page of the stream. Since we don't have page granularity, it would actually have to be done after each individual entry. That should be fine though, we'll just set a variable a bit more often.